matrix-core/ace/config.tdl

```diff
@@ -27,7 +27,7 @@ handle-type               := h.
 list-type                 := list.
 cons-type                 := cons.
 null-type                 := null.
-diff-list-type            := diff-list.
+diff-list-type            := append-list.
```

Comment:
So, this one, I don't know why it is needed but I think Woodley said it was. I will try to dig something up.

Reply:
@emilymbender do you remember why this change was needed? I cannot find anything in my email :/

Reply:
I think this is necessary for generation, because ACE initialises the `CONT` lists using this type.  Unifying `diff-list` and `append-list` will fail.  (And allowing a common subtype might lead to weird bugs.)

Reply:
If this is something that ACE is looking for, is it in fact something we should be changing in a config file instead?

Reply:
Oh, I thought this *was* the config file.

Reply:
Ah, indeed. (Sorry, I'm still really struggling with the github interface.) Then this makes sense: it's telling ACE what to treat as a "diff-list" for its internal purposes.

Reply:
Ok so, as I understand, append lists are *replacing* diff lists rather than being *in addition* to them. So one could perhaps even use the `<! ... !>` syntax to describe append lists?

Also, do we need to similarly update the LKB's `*diff-list-type*` in `lkb/globals.lsp` and PET's `special-name-difflist` in `pet/flop.set`?

Reply:
don't know! @guyemerson @emilymbender ?

Reply:
When I raised the question of extending TDL syntax for append lists (at the summit last year), it was pointed out that the `<! foo !>` syntax is particularly helpful because writing `[ LIST < foo . #last >, LAST #last ]` is tedious.  But with append-lists, `[ LIST < foo > ]` isn't bad.  So there's no strong motivation for special TDL syntax.

For backward compatibility, I wouldn't want to re-define the `<! !>` syntax.

I don't understand the config files, so I can't comment on that.  My previous comment was just remembering what was said in an email.

Reply:
My point about the syntax is not about extending TDL for append lists, but the side effects of changing ACE's `diff-list-type`. If someone sets `diff-list-type := append-list.` *and* has `<! ... !>` in their TDL, then it would instantiate an append list with the coindexed LAST feature, I think. If so, this seems like it would cause hard-to-track-down bugs.

Reply:
Ah, I see what you mean.  Having now read http://moin.delph-in.net/TdlRfc#Feature_interpretation_of_lists , I think I now understand this section of the config.  On closer inspection, I see that the grammar still has: `append-diff-list := append-list & diff-list.`, which is only needed as a workaround for generation.  So I'm not sure the above change to the config is actually necessary.

Reply:
As for a longer-term solution (avoiding `append-diff-list := append-list & diff-list.`), we would need ACE and other systems to avoid non-configurable uses of diff-lists.

Reply:
> I see that the grammar still has: `append-diff-list := append-list & diff-list.`

Which grammar?

Reply:
> Which grammar?

I think he's referring to matrix-core, and by extension any grammar customized with it.

Can you check whether generation works *without* the `ace/config.tdl` change above? If so, maybe it's not necessary, as Guy said.

Reply:
OK, sorry, this is embarrassing, but what's the simplest way to test generation? I can't figure it out (I did look in the wiki). The one way I know about involves setting up MT on the VM but I am assuming that's not necessary?

Reply:
(Also, the reason I put it there is because Emily couldn't do something and Woodley suggested the change :) But yeah, I am happy to check again if I know what to check exactly).

Reply:
Oh, I finally found the email thread. Here's what Emily first observed:

"A while back, Guy made a version of matrix.tdl that uses append-lists instead of diff-lists for RELS and HCONS. Olga has incorporated this into a version of the customization system that we're using in 567 this year. Alas, there's something that ace is expecting that I can't figure out how to fix via config.tdl. The attached mini-English grammar and the sentence "I sleep" provides a test case. When I try generating from the resulting MRS, I get the following error:

ubuntu@UbuntuLKB:~/Downloads/eng$ ace -g eng.dat -Tf1 | ace -g eng.dat -e
I sleep
NOTE: 1 readings, added 15 / 7 edges to chart (6 fully instantiated, 3 actives used, 3 passives used) RAM: 95k
type inference:  no glb for `append-list' and `diff-list'
  LIST appropriate to list-wrapper
  APPEND appropriate to append-list
  LAST appropriate to diff-list
ace: lexicon.c:69: lexical_dg: Assertion `wellform_result == 0' failed.
^C
Aborted (core dumped)

In order to get this far, I changed the value of diff-list-type form diff-list to append-list, but don't see what else I should be changing. Is this in fact something that can be solved via config files, or is some deeper change to ace required?"


And here's what Woodley said:

"What's going on here is that diff-lists for RELS is an assumption built into the generator.  ACE builds a diff-list of the EP specifications contained in the input MRS that align to a particular grammar sign being instantiated, and unifies it in before putting the sign into the chart.  Since ACE built a diff-list but the grammar specified an append-list, a collision occurs.

I added the following line to matrix.tdl and successfully parsed and generated a couple of simple sentences, but I don't have enough context on the implementation of the grammar to know whether this might have unintended side effects:

append-diff-list := append-list & diff-list.

Maybe Olga and Guy could comment on whether they expect the non-unifiability of diff-list and append-list to ever be something the grammar would rely on?  I can't quite imagine when that could possibly happen, especially since the supertype list-wrapper is never explicitly made the appropriate type of any feature.

Possibly a better long-term solution would be for ACE to have a config option to apply these sign specializations as append-lists instead of diff-lists, but since this patch seems to work for the moment and using append-lists for RELS is not exactly mainstream yet, I think I will hold off on that until there's a clearer need."

Reply:
Ok, thanks for the context. Woodley doesn't actually say that the config change is necessary, but the matrix.tdl one currently is. So I tried 4 variations with the `web/sample-choices/mini-english` choices file, and tried generating with:

    $ echo "the dog sleeps" | ace -g eng.dat | ace -g eng.dat -e

1. Without config change, without `append-diff-list` -> crashed
2. With config change, without `append-diff-list` -> crashed
3. Without config change, with `append-diff-list` -> succeeded
4. With config change, with `append-diff-list` -> succeeded

It seems like it makes no difference, for this test, whether the config change is made or not. Maybe we should just leave it out altogether for now?

Reply:
Thanks for checking this.  Yes, I think it makes sense to leave out this change.

Reply:
Yeah, I see now how  I probably made this change because it seemed that Emily needed it where in fact she needed a different one. 
Great! Done.